### ISSUE TITLE
API Version 3.1 doesn't work with Resources other than Send

### ISSUE THREAD

**Issue by XavAM** [mailjet/mailjet-apiv3-dotnet#37]
It took me quite a while to understand that the endpoint v3.1 only works for sending messages, and get a non explicit "Not Found" response when trying other endpoints (like retrieving contacts, messsages, and so on). 

It should be either really detailed into the doc, either prevented by the wrapper. 

Example : 

>            var  client = new MailjetClient(m_configuration.ApiKey, m_configuration.ApiSecret)
>            {
>                Version = ApiVersion.V3_1,
>            };
>            MailjetRequest request = new MailjetRequest
>          {
>               Resource = Message.Resource,
>           };
>          MailjetResponse response = await client.GetAsync(request); //gets a 404 without any detail

This **never** works. And I am quite annoyed to have spent two hours figuring out why :-(

**Comment by djrhodes** [comment]
It should be handled by the client wrapper. I configure my Mailjet client when I register it in the DI container and specify the api version there. Send works fine but when I do a call to retrieve template details, I have to try a safe cast to the concrete Mailjet client and if successful then set the ApiVersion to v3. This should be handled in the client.

**Comment by XavAM** [comment]
I don't understand your answer. What you are saying is that you specify the version when register in the DI container, then try to retrieve tempalte details, try to cast it,  and set the version again? 

This doesn't really make sense. 

What's the point in exposing a method which can never succeed ? It woud make more sense to have different wrappers, for different versions, exposing only relevant methods. 
 
And, just for a reminder, we pay a significant amount of money to Mailjet for a service. I think we deserve a little bit more consideration than just a "this should be handled by the client" answer type...

**Comment by djrhodes** [comment]
@XavAM I’m on your side - when I say the client should handle it I mean the Mailjet client. The cast and reset version thing is a ridiculous workaround that I shouldn’t have to do. The .Net mailjet client is somewhat confusing and incomplete (the constants weren’t even updated to include the v3_1 properties). That said it is OSS, so rather than whinging and implementing hacky workarounds we could always roll our sleeves fork the code and submit a PR 😆

**Comment by burakdobur** [comment]
I also +1 XavAM, we are paying our money to Mailjet to save our resources/time. Not every company have resources or policies to improve missing features that should be there at first. Please dedicate "a single devloper" to complete missing features and refactor your .NET client Mailjet people.

**Comment by Emths** [comment]
> 
> 
> It should be handled by the client wrapper. I configure my Mailjet client when I register it in the DI container and specify the api version there. Send works fine but when I do a call to retrieve template details, I have to try a safe cast to the concrete Mailjet client and if successful then set the ApiVersion to v3. This should be handled in the client.

Hello all,
@djrhodes : Sorry, but I don't understand what I have to do to make it work... Could you please post a code sample to explain how to do it?
Thanks a lot!

**Comment by djrhodes** [comment]
Sorry, I don’t have the code to hand but basically you can specify the api version when you new up an instance of the Mailjet client for the new Send API specify 3.1 for everything else e.g templates specify 3.

Hope this helps,
John

**Comment by Emths** [comment]
@djrhodes : ok, thanks! I will try it as soon as possible! :-)

**Comment by goacoustic** [comment]
@XavAM Hi, thanks for raising this question!

I've added a possibility to use strongly typed models for sending transactional emails, as this is one of the most popular features.
You can check how it looks like [here](https://github.com/mailjet/mailjet-apiv3-dotnet/blob/e150ab97c067a426478bb4a2c18c757b1c4cb52b/Mailjet.Tests/Integration/SendTransactionalEmailIntegrationTests.cs#L40)

If you have any other suggestions - please, open a separate issue, so we can track that faster. :)

**Comment by actionbin** [comment]
@ goacoustic I currently switched to using MailJet.Api version 2.0.1. It appears that the Send.Resource is still pointing to version 3 instead of 3.1. 

@XavAM I currently have the work around below...
            var sendResource = new ResourceInfo("send", null, ApiVersion.V3_1, ResourceType.Send);

            MailjetRequest request =
                new MailjetRequest { Resource = sendResource }
                    .Property(Send.Messages, new JArray {
                        new JObject {
                            { "From", sender },
                            { "To", recipient },
                            { "Subject", mail.Subject },
                            { "TextPart", mail.TextPart },
                            { "HTMLPart", mail.HtmlPart },
                            { "CustomID", mail.MailId }
                        }
                    });

**Comment by goacoustic** [comment]
@actionbin Hi!

There is SendV31 resource available, which basically does the same. 
Also, the latest version has a builder that uses v3.1 under the hood and makes the code more readable.

```c#
var email = new TransactionalEmailBuilder()
                .WithFrom(new SendContact("from@test.com"))
                .WithSubject("Test subject")
                .WithHtmlPart("<h1>Header</h1>")
                .WithTo(new SendContact("to@test.com"))
                .Build();

var response = await _client.SendTransactionalEmailAsync(email);
```

**Comment by kimmymonassar** [comment]
Hey!

This has been closed for a year... Can we get an updated API doc over at dev.mailjet.com?
[Send API docs](https://dev.mailjet.com/email/guides/send-api-v31/) still specifies the Version = ApiVersion_V3 under C# examples if the parameter doesn't even exist anymore? 

Would save a lot of trial and error to have updated examples in the actual API documentation
